### ISSUE TITLE
get_bands interpolation

### ISSUE THREAD

**Issue by Simonsays095** [pyscf/pyscf#1063]
I've been getting some weird results when computing the band structure with k-points that aren't directly used in SCF. Specifically, I've done a few calculations with 9 k-points in SCF, and then attempting to compute a "smoother" band structure with 50 k-points. Since the get_bands function takes `kpts_band` and `kpts` separately, I assume this is valid to do. Below is the image of my band structure, showing the odd "bumps" between k-points used directly in the SCF (each cusp in the occupied orbitals is one such k-point).
![image](https://user-images.githubusercontent.com/47613665/134948153-03acf7d9-6bcc-4fd1-8a3e-7f44d80d9e47.png)

My first question: what interpolation technique is used? I've looked through the code and can't haven't found a reference or code snippet I can understand well-enough to glean what underlying theory is used.

And secondly: am I doing it right? It seems incredibly simple to plot the band structure, but in case I'm missing something, this is this code I use to collect band energies and plot them:
```py
# ... ^^ set up unit cell and perform HF or DFT calculation

band_lines = cell.make_kpts([50,1,1]) # 1D system
band_e, _ = mf.get_bands(band_lines)
band_e = np.asarray(band_e)

plt.plot(band_lines[:,0], band_e) # Plot line along kx-axis
plt.show()
```

Thanks for your help!

**Comment by tberkel** [comment]
There is no interpolation in the conventional sense. The get_bands function
simply uses the converged density matrix from an SCF calculation to build
and diagonalize the Fock/KS matrix in a one-shot manner. This is similar to
the difference between "scf" and "nscf" calculations in Quantum Espresso,
if you have used that code before.

Regarding your bumps: are you doing calculations with nonlocal exchange (HF
or hybrid DFT)? If you haven't already, please see the User Guide
documentation and code example (for pure DFT):
https://pyscf.org/user/pbc/scf.html
https://pyscf.org/user/pbc/scf.html#band-structure-calculations
https://github.com/pyscf/pyscf/blob/master/examples/pbc/09-band_ase.py

**Comment by Simonsays095** [comment]
The links you provided are very useful, and I haven't previously seen the band structure calculation section in the user guide. I am indeed using a hybrid functional (B3PW91), so I'll try doing a calculation with `exxdiv='vcut_sph'` set and see if that helps to fix the bumps. Will this parameter affect the results of the SCF calculation?

Also, the example script you sent https://github.com/pyscf/pyscf/blob/master/examples/pbc/09-band_ase.py has an example where the band structure is computed from a gamma-point calculation. Without doing another SCF calculation, I don't understand how `get_bands` computes the density matrix for the additional k-points (I have not used quantum espresso before, let alone the nscf method). The closest I can think of is to:
1. Construct hcore for the larger k-point basis.
2. Construct veff using the smaller k-point basis's density matrix, making an assumption about the form of the density matrix in "new" k-points.
3. Construct and diagonalize the new Fock matrix from hcore and veff, to get the MO energies/MO coefficients and a use these to get the new density matrix.

But even with this method, there is an assumption about what the density matrix looks like for the k-points not in the original calculation. Looking through the code, I'm pretty sure this isn't what's actually happening, it's just the only method I can think of. I've also heard of a method called the "tetrahedron" method: is this used in PySCF, or am I off-track with both methods?

**Comment by tberkel** [comment]
Yes, the different choices for `exxdiv` will give different results but
they all agree in the thermodynamic limit of infinite k-point sampling.

Regarding band structure: first, recall that hcore is only defined by a
single k-point (and not a mesh of k-points). If you neglected interactions,
you could build and diagonalize h(k) at every k-point of interest in your
band structure (with no approximation). Similarly, veff (and J and K) only
depends explicitly on a single k-point, but it depends implicitly on other
k-points through the density or density matrix. One way to see why the kpts
used in the SCF mesh don't have to be the same as the kpt at which you want
the band structure is to think of rho(r) or rho(r,r') in real space, which
are obtained approximately through some Brillouin zone integral. Some of
this might become clearer by reading the SCF section of our older paper:
https://arxiv.org/pdf/1701.04832.pdf -- there is no reason that the set of
k's that are used in (13) to approximate the DM need to be the same as the
k's at which you evaluate the J and K (and Fock) matrices in (10) and (11).
So (13) can be done using only a single k-point (the Gamma point),
providing some approximate density matrix, from which the J and K matrices
in (10) and (11) can be calculated for *any* k-point in the Brillouin zone
(and then Fock is diagonalized to get the band structure).

**Comment by Simonsays095** [comment]
Hi Tim,

Thank you very much for your detailed explanation and the link to the previous paper, it's much appreciated and very helpful. I think I understand the idea behind the reconstruction of the 1-RDM for any k-point, without making additional approximations. You construct the actual real-space density matrix on a grid using the 1-RDM, and then extract J/K for other k-points for an arbitrary set of k-points in equations (10) and (11).

As for exxdiv: I tried running a calculation with `exxdiv='vcut_sph'`, but I got the error
```bash
RuntimeError: GDF does not support exxdiv vcut_sph
```
This is raised from https://github.com/pyscf/pyscf/blob/09834c8f5a4f5320cdde29d285b6ccd89b3263e0/pyscf/pbc/df/df_jk.py#L154
when trying to compute the K matrix for several k-points, where it looks like either no exxdiv or ewald are the only valid options.

**Comment by tberkel** [comment]
Yes, unfortunately, GDF only supports those two at this time. One day, we
should be able to have `vcut_sph` supported. Alternatively, you can try
FFTDF, which supports all of the `exxdiv` options, but might have other
issues depending on your system. A faster FFTDF is also in development and
should be released soon. Lastly, you can re-do the SCF with a kpt mesh that
is centered on each point in your band structure path (as I mentioned
originally) -- in this case, GDF can be used with `ewald` and you will
still get a smooth band structure. The downside of course is that you have
to do many SCF calculations.

**Comment by Simonsays095** [comment]
FFTDF only supports unit cells that are periodic in 2 or 3 dimensions, right? Then 2D or 3D systems have the option to use FFTDF/exxdiv, but to compute the band structure for a 1D chain calculation using a hybrid functional, the only option (currently) is to perform a different SCF calculation for each kpt. Each of these SCF calculations can be done with a single kpt, specified using the scaled_center option when generating the kpt (to summarize your suggestion). Since this can be parallelized, it shouldn't take too long as far as wall time, but it will take a lot more computation time in total.

Otherwise, I can use GDF and the ewald exxdiv parameter if I switch to a non-hybrid functional? Earlier, you asked if I was doing HF or using a hybrid functional at the same time, because they both have nonlocal response - will I face the same problem when doing a typical HF calculation? If so, what options do I have to do get the band structure from a typical HF calculation on a 1D chain?

**Comment by tberkel** [comment]
I haven't checked the options for low-dimensional calculations with cell.dimension recently, but you may be right. However, a 1D system can always be done with cell.dimension=3 by using a large enough vacuum. In that case, you can surely use FFTDF, enabling the choice of `exxdiv='vcut_sph'` and normal use of the `get_bands()` function.

If you can only use GDF and you are doing HF or hybrid DFT and you want the band structure, then you are correct that you have to perform the SCF at each k-point in your band path, using `scaled_center`. However, note that this does not mean that you only have to use one k-point in your mesh. Doing so will incur some sampling/finite-size error. Ideally you should use increasingly large kpt meshes until convergence of the band structure (but these of course become more expensive). Here is some code (h/t my student Ethan Vo) to demonstrate the idea:

```
npoints_in_path = 10 # number of points in the band path, SCF for each one
nks = [3,3,3] # density of BZ sampling for each SCF -- should be increased for convergence
bandpath = get_bandpath('GX', ase_atom.cell, npoints=npoints_in_path) # helper function from ASE
for kcenter in bandpath.kpts:
    # make a new kpt mesh and do a new SCF
    kpts = cell.make_kpts(nks, scaled_center=kcenter)
    mymf = scf.KRHF(cell, kpts=kpts, exxdiv='ewald').density_fit()
    # ... mo energies of the *first* kpt will be the band energies at the kpt 'kcenter'
```

**Comment by tberkel** [comment]
One last point: FFTDF can't be practically used without a pseudopotential.
But otherwise, this all looks correct to me. Thanks for summarizing. Feel
free to reopen if anything doesn't work as expected.
